llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp

```diff
@@ -6010,8 +6010,94 @@ bool AMDGPUTargetLowering::shouldSinkOperands(
     if (any_of(Ops, [&](Use *U) { return U->get() == Op.get(); }))
       continue;
 
+    // This helps using several modifiers (abs and neg) more often.
```

Comment:
Looking for all the free operations is already in TargetTransformInfo. I think all of these IR hooks in TargetLowering for CodeGenPrepare are for historical reasons. Can you try just moving it there, and checking for free costs?